### PR TITLE
Dev pm 5372; event localizer refactor

### PR DESCRIPTION
**Pull request by PeterMeisrimelModelon**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -460,17 +460,22 @@ def check_LAPACK(self):
...
-        ext_list = cythonize(["assimulo"+os.path.sep+"*.pyx"], 
-                             include_path=[".","assimulo"])
-
-        #Cythonize Solvers
+        # Cythonize main modules 
+        ext_list = cythonize([os.path.join("assimulo", "explicit_ode.pyx")], 
+                             include_path=[".", "assimulo"], force = True)
+        ext_list[-1].include_dirs += ["assimulo", self.incdirs]
+        ext_list[-1].sources += [os.path.join("assimulo", "ode_event_locator.c")]
+        ext_list[-1].name = "assimulo.explicit_ode"
```
Comment (chria) [inline setup.py:468]: Do you know why we need a name here, while not for the others?

File: src/explicit_ode.pyx [hunk 2]
```diff
@@ -23,13 +23,47 @@ import sys
...
+cdef void c2py_d(N.ndarray[double, ndim=1, mode='c'] dest, double* source, int dim):
+    """Copy (double *) C vector to 1D numpy array."""
+    memcpy(N.PyArray_DATA(dest), source, dim*sizeof(double))
+
+cdef int callback_event(int n_y, int n_g, double t, double* y_in, double* g_out, void* f_event_EXT):
+    """Event indicator callback function to event_locator.c"""
+    cdef N.ndarray[double, ndim=1, mode="c"]y_py = N.empty(n_y, dtype = N.double)
+    c2py_d(y_py, y_in, n_y)
+    g_high = (<object>f_event_EXT)(t, y_py)
+    # print(f"callback event, t = {t}, g_high = {g_high}, type = {type(g_high)}")
```
Comment (chria) [inline src/explicit_ode.pyx:56]: Remove?

File: src/explicit_ode.pyx [hunk 3]
```diff
@@ -23,13 +23,47 @@ import sys
...
+    py2c_d(g_out, g_high, n_g)
+    return 0
+
+cdef int callback_interp(int n, double t, double* y_out, void* f_interp_EXT):
+    """Interpolation callback function to event_locator.c"""
+    y_interp = (<object>f_interp_EXT)(t)
+    # print(f"Callback interpolation, t = {t}, y = {y_interp}, type =  {type(y_interp)}")
```
Comment (chria) [inline src/explicit_ode.pyx:63]: Remove?

File: src/solvers/euler.pyx [hunk 4]
```diff
@@ -103,6 +103,7 @@ cdef class ImplicitEuler(Explicit_ODE):
             self.event_func = event_func
             self._event_info = N.array([0] * self.problem_info["dimRoot"]) 
             self.g_old = self.event_func(self.t, self.y)
+            self.statistics["nstatefcns"] += 1
```
Comment (chria) [inline src/solvers/euler.pyx:106]: Should it really by nstatefcns?

File: src/solvers/euler.pyx [hunk 5]
```diff
@@ -535,6 +536,7 @@ cdef class ExplicitEuler(Explicit_ODE):
...
```
Comment (chria) [inline src/solvers/euler.pyx:539]: Same here (and below), is this correct?

**Review by chria — COMMENTED** [review]
Just minor comments, looks overall good!

File: src/solvers/euler.pyx (hunk 4 above)
Comment (PeterMeisrimelModelon) [reply src/solvers/euler.pyx:106]: In ode.pyx it is defined by

self.statistics.add_key("nstatefcns", "**Number of state function evaluations**")

So, yes.

File: src/solvers/euler.pyx (hunk 5 above)
Comment (PeterMeisrimelModelon) [reply src/solvers/euler.pyx:539]: See previous comment.

File: src/explicit_ode.pyx (hunk 3 above)
Comment (PeterMeisrimelModelon) [reply src/explicit_ode.pyx:63]: Removed.

File: setup.py (hunk 1 above)
Comment (PeterMeisrimelModelon) [reply setup.py]: Actually, after testing it again, it appears redundant. Removed it.

**Pull request merged**